include/pybind11/typing.h

```diff
@@ -0,0 +1,94 @@
+/*
+    pybind11/typing.h: Convenience wrapper classes for basic Python types
+    with more explicit annotations.
+
+    Copyright (c) 2016 Wenzel Jakob <wenzel.jakob@epfl.ch>
+
+    All rights reserved. Use of this source code is governed by a
+    BSD-style license that can be found in the LICENSE file.
+*/
+
+#pragma once
+
+#include "detail/common.h"
+#include "cast.h"
+#include "pytypes.h"
+
+PYBIND11_NAMESPACE_BEGIN(PYBIND11_NAMESPACE)
+
+/*
+    The following types can be used to direct pybind11-generated docstrings
+    to have have more explicit types (e.g., `List[str]` instead of `list`).
+    Just use these in place of existing types.
+
+    There is no additional enforcement of types at runtime.
+*/
+
+template <typename... Types>
+class Tuple : public tuple {
+    using tuple::tuple;
+};
+
+template <typename K, typename V>
+class Dict : public dict {
+    using dict::dict;
+};
+
+template <typename T>
+class List : public list {
```

Comment:
Could this be `template <typename... Types>` like or `Tuple`?

Reply:
IMO, `list[float | str]` is fine / best, that's valid for older versions of Python as long as you only use it for typing and not at runtime.

Reply:
Also, the equivalent for `tuple` is `tuple[float | str, ...]`. `tuple[float, str]` is a size 2 tuple with a float as the first item, and a str as the second.